File: frontend/src/runtime/three/SeededRandom.ts
export class SeededRandom {
  private state: number;

  constructor(seed: number) {
    this.state = seed >>> 0 || 0x12345678;
  }

  next(): number {
    this.state = (1664525 * this.state + 1013904223) >>> 0;
    return this.state / 0x100000000;
  }

  range(min: number, max: number): number {
    return min + (max - min) * this.next();
  }

  centered(amount = 1): number {
    return (this.next() * 2 - 1) * amount;
  }
}

File: Metin2AssetStudioCpp/src/runtime/SeededRandom.cpp
#include "SeededRandom.h"

SeededRandom::SeededRandom(uint64_t seed) : state(seed) {}

double SeededRandom::next() {
    state = (1664525ULL * state + 1013904223ULL) & 0xFFFFFFFF;
    return state / 4294967296.0;
}

double SeededRandom::range(double min, double max) {
    return min + next() * (max - min);
}

double SeededRandom::centered(double amount) {
    return next() * 2.0 * amount - amount;
}
